File: tools/http/test-client/js/src/contract.js
// Copyright The OpenTelemetry Authors
// SPDX-License-Identifier: Apache-2.0

"use strict";

/** The HTTP conformance exchanges supplied by the runner as JSON. */

/** Every route answers JSON, so a scenario has one content type rather than a rule per route. */
const CONTENT_TYPE = "application/json";

/**
 * Fixed rather than the HTTP library's default, so a server scenario sees the
 * same client whichever language sent the requests.
 */
const USER_AGENT = "otel-http-conformance/1";
const ACTION_VARIABLE = "OTEL_CONFORMANCE_SCENARIO_ACTION";
const ACTIONS_VARIABLE = "OTEL_CONFORMANCE_SCENARIO_ACTIONS";

function parseJson(raw, variable) {
  try {
    return JSON.parse(raw);
  } catch (error) {
    throw new Error(`${variable} contains malformed JSON: ${error.message}`, {
      cause: error,
    });
  }
}

function checkKeys(value, allowed, where) {
  const unknown = Object.keys(value).filter((key) => !allowed.has(key));
  if (unknown.length) {
    throw new Error(
      `${where} has unknown field(s): ${unknown.sort().join(", ")}`,
    );
  }
}

function actionExchange(value, variable, readiness) {
  const where = `${variable} action`;
  if (value === null || typeof value !== "object" || Array.isArray(value)) {
    throw new Error(`${where} must be a JSON object`);
  }
  checkKeys(value, new Set(["request", "response"]), where);
  if (!("request" in value) || !("response" in value)) {
    throw new Error(`${where} requires request and response objects`);
  }
  for (const field of ["request", "response"]) {
    if (
      value[field] === null ||
      typeof value[field] !== "object" ||
      Array.isArray(value[field])
    ) {
      throw new Error(`${where}.${field} must be a JSON object`);
    }
  }

  const request = value.request;
  const response = value.response;
  checkKeys(request, new Set(["method", "path", "body"]), `${where}.request`);
  checkKeys(response, new Set(["status", "body"]), `${where}.response`);
  if (typeof request.method !== "string" || request.method.length === 0) {
    throw new Error(`${where}.request.method must be a non-empty string`);
  }
  if (typeof request.path !== "string" || !request.path.startsWith("/")) {
    throw new Error(`${where}.request.path must start with '/'`);
  }
  if (
    request.body !== undefined &&
    request.body !== null &&
    typeof request.body !== "string"
  ) {
    throw new Error(`${where}.request.body must be a string`);
  }
  if (
    !Number.isInteger(response.status) ||
    response.status < 100 ||
    response.status > 599
  ) {
    throw new Error(`${where}.response.status must be an HTTP status`);
  }
  if (typeof response.body !== "string") {
    throw new Error(`${where}.response.body must be a string`);
  }
  return Object.freeze({
    method: request.method,
    path: request.path,
    body: request.body ?? null,
    status: response.status,
    responseBody: response.body,
    readiness,
    description: readiness ? "runner readiness action" : "runner action",
  });
}

/** Every exchange the runner supplied, including readiness, in order. */
function exchanges(raw = process.env[ACTIONS_VARIABLE]) {
  if (raw === undefined) {
    throw new Error(`${ACTIONS_VARIABLE} is not set`);
  }
  const actions = parseJson(raw, ACTIONS_VARIABLE);
  if (!Array.isArray(actions) || actions.length === 0) {
    throw new Error(
      `${ACTIONS_VARIABLE} must be a non-empty JSON array of actions`,
    );
  }
  return Object.freeze(
    actions.map((action, index) =>
      actionExchange(action, `${ACTIONS_VARIABLE}[${index}]`, index === 0),
    ),
  );
}

/** The measured requests supplied by the runner. */
function requests(raw = process.env[ACTIONS_VARIABLE]) {
  return Object.freeze(exchanges(raw).slice(1));
}

/** The one request selected by the runner. */
function scenarioRequest(raw = process.env[ACTION_VARIABLE]) {
  if (raw === undefined) {
    throw new Error(`${ACTION_VARIABLE} is not set`);
  }
  return actionExchange(
    parseJson(raw, ACTION_VARIABLE),
    ACTION_VARIABLE,
    false,
  );
}

function withoutQuery(target) {
  return target.split("?", 1)[0];
}

/** The exchange answering `method path`, if the runner supplied one. */
function exchangeFor(method, target) {
  const concrete = withoutQuery(target);
  return (
    exchanges().find(
      (exchange) =>
        exchange.method === method && withoutQuery(exchange.path) === concrete,
    ) ?? null
  );
}

/** An exchange's response body with the request body inserted. */
function renderResponseBody(exchange, requestBody) {
  return exchange.responseBody.replace("${requestBody}", () =>
    requestBody ? requestBody : "{}",
  );
}

module.exports = {
  ACTIONS_VARIABLE,
  ACTION_VARIABLE,
  CONTENT_TYPE,
  USER_AGENT,
  exchangeFor,
  exchanges,
  renderResponseBody,
  requests,
  scenarioRequest,
};

File: tools/http/test-client/go/contract.go
// Copyright The OpenTelemetry Authors
// SPDX-License-Identifier: Apache-2.0

// Package httpcontract decodes the HTTP conformance actions supplied by the runner.
package httpcontract

import (
	"encoding/json"
	"fmt"
	"io"
	"os"
	"strings"
	"sync"
)

const ContentType = "application/json"
const UserAgent = "otel-http-conformance/1"
const ActionVariable = "OTEL_CONFORMANCE_SCENARIO_ACTION"
const ActionsVariable = "OTEL_CONFORMANCE_SCENARIO_ACTIONS"

type Exchange struct {
	Method       string
	Path         string
	Body         string
	Status       int
	ResponseBody string
	Readiness    bool
	Description  string
}

func (e Exchange) RenderResponseBody(requestBody string) string {
	if requestBody == "" {
		requestBody = "{}"
	}
	return strings.ReplaceAll(e.ResponseBody, "${requestBody}", requestBody)
}

type Response struct {
	StatusCode int
	Body       string
}

type Error struct {
	message string
	cause   error
}

func (e *Error) Error() string { return e.message }
func (e *Error) Unwrap() error { return e.cause }

func contractError(format string, arguments ...any) error {
	return &Error{message: fmt.Sprintf(format, arguments...)}
}

type action struct {
	Request  request  `json:"request"`
	Response response `json:"response"`
}

type request struct {
	Method string  `json:"method"`
	Path   string  `json:"path"`
	Body   *string `json:"body,omitempty"`
}

type response struct {
	Status int     `json:"status"`
	Body   *string `json:"body"`
}

var actionCache struct {
	sync.Mutex
	raw       string
	exchanges []Exchange
}

func Exchanges() ([]Exchange, error) {
	raw, ok := os.LookupEnv(ActionsVariable)
	if !ok {
		return nil, fmt.Errorf("%s is not set", ActionsVariable)
	}
	actionCache.Lock()
	defer actionCache.Unlock()
	if actionCache.exchanges != nil && actionCache.raw == raw {
		return actionCache.exchanges, nil
	}

	var actions []action
	if err := decodeJSON(raw, ActionsVariable, &actions); err != nil {
		return nil, err
	}
	if len(actions) == 0 {
		return nil, fmt.Errorf("%s must be a non-empty JSON array of actions", ActionsVariable)
	}
	exchanges := make([]Exchange, len(actions))
	for index, action := range actions {
		exchange, err := action.exchange(index == 0, fmt.Sprintf("%s[%d]", ActionsVariable, index))
		if err != nil {
			return nil, err
		}
		exchanges[index] = exchange
	}
	actionCache.raw = raw
	actionCache.exchanges = exchanges
	return exchanges, nil
}

func Requests() ([]Exchange, error) {
	exchanges, err := Exchanges()
	if err != nil {
		return nil, err
	}
	return exchanges[1:], nil
}

func ScenarioRequest() (Exchange, error) {
	raw, ok := os.LookupEnv(ActionVariable)
	if !ok {
		return Exchange{}, fmt.Errorf("%s is not set", ActionVariable)
	}
	var selected action
	if err := decodeJSON(raw, ActionVariable, &selected); err != nil {
		return Exchange{}, err
	}
	return selected.exchange(false, ActionVariable)
}

func Lookup(method, path string) (Exchange, bool, error) {
	exchanges, err := Exchanges()
	if err != nil {
		return Exchange{}, false, err
	}
	for _, exchange := range exchanges {
		if exchange.Method == method && exchange.Path == path {
			return exchange, true, nil
		}
	}
	pathWithoutQuery := withoutQuery(path)
	for _, exchange := range exchanges {
		if exchange.Method == method && withoutQuery(exchange.Path) == pathWithoutQuery {
			return exchange, true, nil
		}
	}
	return Exchange{}, false, nil
}

func (a action) exchange(readiness bool, where string) (Exchange, error) {
	if a.Request.Method == "" {
		return Exchange{}, fmt.Errorf("%s.request.method must be a non-empty string", where)
	}
	if !strings.HasPrefix(a.Request.Path, "/") {
		return Exchange{}, fmt.Errorf("%s.request.path must start with '/'", where)
	}
	if a.Response.Status < 100 || a.Response.Status > 599 {
		return Exchange{}, fmt.Errorf("%s.response.status must be an HTTP status", where)
	}
	if a.Response.Body == nil {
		return Exchange{}, fmt.Errorf("%s.response.body must be a string", where)
	}
	body := ""
	if a.Request.Body != nil {
		body = *a.Request.Body
	}
	description := "runner action"
	if readiness {
		description = "runner readiness action"
	}
	return Exchange{
		Method:       a.Request.Method,
		Path:         a.Request.Path,
		Body:         body,
		Status:       a.Response.Status,
		ResponseBody: *a.Response.Body,
		Readiness:    readiness,
		Description:  description,
	}, nil
}

func decodeJSON(raw, variable string, target any) error {
	decoder := json.NewDecoder(strings.NewReader(raw))
	decoder.DisallowUnknownFields()
	if err := decoder.Decode(target); err != nil {
		return fmt.Errorf("%s contains malformed JSON: %w", variable, err)
	}
	if err := decoder.Decode(new(any)); err != io.EOF {
		if err == nil {
			return fmt.Errorf("%s contains more than one JSON value", variable)
		}
		return fmt.Errorf("%s contains malformed JSON: %w", variable, err)
	}
	return nil
}

func withoutQuery(path string) string {
	if query := strings.IndexByte(path, '?'); query != -1 {
		return path[:query]
	}
	return path
}

func parse(body string) (any, error) {
	var parsed any
	if err := json.Unmarshal([]byte(body), &parsed); err != nil {
		return nil, &Error{
			message: fmt.Sprintf("not JSON: %s: %v", abbreviate(body), err),
			cause:   err,
		}
	}
	return parsed, nil
}
